File: Data Structures & Algorithms/climbing-stairs/submission-0.py
class Solution:
    def climbStairs(self, n: int) -> int:
        return self.dfs(1, n) + self.dfs(2, n)
    
    def dfs(self, length : int, n : int) -> int:
        if length > n:
            return 0
        if length == n:
            return 1
        res = 0
        for i in range(1, 3):
            res += self.dfs(length + i, n)
        return res

File: Data Structures & Algorithms/generate-parentheses/submission-0.java
class Solution {
    private boolean valid(String s) {
        int add = 0;
        for(char c: s.toCharArray()) {
            add += c == '(' ? 1 : -1;
            if(add < 0) return false;
        }
        return add == 0;
    }

    private void dfs(String s, List<String> res, int n) {
        if(s.length() == n * 2) {
            if(valid(s)) res.add(s);
            return;
        }
        dfs(s + '(', res, n);
        dfs(s + ')', res, n);
    }

    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        dfs("", res, n);
        return res;
    }
}
